src/NuGet.Core/NuGet.Packaging/Signing/Timestamp/Rfc3161TimestampProvider.cs

```diff
@@ -0,0 +1,198 @@
+        public Task<Signature> TimestampSignatureAsync(TimestampRequest request, ILogger logger, CancellationToken token)
+        {
+            var timestampedSignature = TimestampData(request, logger, token);
+            return Task.FromResult(Signature.Load(timestampedSignature));
+        }
+
+        /// <summary>
+        /// Timestamps data present in the TimestampRequest.
+        /// </summary>
+        public byte[] TimestampData(TimestampRequest request, ILogger logger, CancellationToken token)
+        {
+            token.ThrowIfCancellationRequested();
+
+            if (request == null)
+            {
+                throw new ArgumentNullException(nameof(request));
+            }
+
+            if (logger == null)
+            {
+                throw new ArgumentNullException(nameof(logger));
+            }
+
+            // Get the signatureValue from the signerInfo object
+            using (var signatureNativeCms = NativeCms.Decode(request.SignatureValue, detached: false))
+            {
+                var signatureValueHashByteArray = GetSignatureValueHash(
+                    request.TimestampHashAlgorithm,
+                    signatureNativeCms);
+
+                // Allows us to track the request.
+                var nonce = GenerateNonce();
+                var rfc3161TimestampRequest = new Rfc3161TimestampRequest(
+                    signatureValueHashByteArray,
+                    request.TimestampHashAlgorithm.ConvertToSystemSecurityHashAlgorithmName(),
+                    nonce: nonce,
+                    requestSignerCertificates: true);
+
+                // Request a timestamp
+                // The response status need not be checked here as lower level api will throw if the response is invalid
+                var timestampToken = rfc3161TimestampRequest.SubmitRequest(
+                    _timestamperUrl,
+                    TimeSpan.FromSeconds(_rfc3161RequestTimeoutSeconds));
+
+                // ensure response is for this request
+                ValidateTimestampResponseNonce(nonce, timestampToken);
+                var timestampCms = timestampToken.AsSignedCms();
+
+                var timestampCertChain = GetTimestampCertChain(GetTimestampSignerCertificate(timestampCms));
+                byte[] timestampByteArray;
+
+                using (var timestampNativeCms = NativeCms.Decode(timestampCms.Encode(), detached: false))
+                {
+                    // Insert all the certificates into timestampCms
+                    InsertTimestampCertChainIntoTimestampCms(timestampCms, timestampCertChain, timestampNativeCms);                    
+                    timestampByteArray = timestampCms.Encode();
+                }
+
+                signatureNativeCms.AddTimestamp(timestampByteArray);
+
+                return signatureNativeCms.Encode();
+            }
+        }
+
+        private static X509Chain GetTimestampCertChain(X509Certificate2 timestampSignerCertificate)
```

Comment:
Should just be a generic method in signing utility? It creates a very timestamp specific error, but it seems like it is just getting the cert chain which will be needed for other scenarios I would think

Reply:
There is already a method in signingutility which does certificate validity check - `IsCertificateValid`. This is to wrap around that to have a better error.